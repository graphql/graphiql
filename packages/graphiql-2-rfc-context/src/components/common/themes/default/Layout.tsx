/** @jsx jsx */

/**
 *  Copyright (c) 2020 GraphQL Contributors.
 *
 *  This source code is licensed under the MIT license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 */

import { PropsWithChildren } from 'react';
import { GraphiQLTheme, PanelSize, LayoutPropTypes } from '../types';

import { jsx, SxStyleProp } from 'theme-ui';

const NAV_WIDTH = '6em';
const CONTENT_MIN_WIDTH = '60em';

function sizeInCSSUnits(theme: GraphiQLTheme, size: PanelSize) {
  switch (size) {
    case 'sidebar':
      return '10em';
    case 'aside':
      return '20em';
    default:
      return `calc(100vw - ${theme.space[2] * 3}px - ${NAV_WIDTH})`;
  }
}

type CardPropTypes = PropsWithChildren<{
  size?: PanelSize;
  transparent?: boolean;
  innerSx?: SxStyleProp;
}>;

const Card = ({
  children,
  size,
  transparent = false,
  innerSx,
}: CardPropTypes) => (
  <div
    sx={
      {
        display: 'grid',
        backgroundColor: !transparent ? 'cardBackground' : undefined,
        boxShadow: !transparent && 'card',
        minWidth: size && (theme => sizeInCSSUnits(theme, size)),
        gridTemplate: '100% / 100%',
        ...(innerSx ?? {}),
      } as SxStyleProp
    }>
    {children}
  </div>
);

const gridBase = {
  display: 'grid',
  gridAutoFlow: 'column',
  gridAutoColumns: '1fr',
  gridAutoRows: '100%',
  gap: 3,
};

const Layout = ({ nav, navPanels, session }: LayoutPropTypes) => {
  const hasNavPanels = (navPanels && navPanels?.length > 0) || false;
  return (
    <main
      sx={{
        ...gridBase,
        padding: 3,
        gridTemplate: hasNavPanels
          ? `'nav panels session' 100% / ${NAV_WIDTH} min-content minmax(${CONTENT_MIN_WIDTH}, 1fr)`
          : `'nav session' 100% / ${NAV_WIDTH} minmax(${CONTENT_MIN_WIDTH}, 1fr)`,
        height: '100%',
      }}>
      {nav && (
        <Card innerSx={{ gridArea: 'nav' }} transparent>
          {nav}
        </Card>
      )}
      {hasNavPanels && (
        <div
          sx={{
            gridArea: 'panels',
            ...gridBase,
          }}>
          {navPanels!.map(({ component, key, size }) => (
            <Card key={key} size={size}>
              {component}
            </Card>
          ))}
        </div>
      )}
      {session && (
        <div
          sx={{
            ...gridBase,
            gridArea: 'session',
            gridAutoRows: '1fr',
            gridTemplateAreas: `'input response' 'console console'`,
          }}>
          <Card innerSx={{ gridArea: 'input' }}>{session.input}</Card>
          <Card innerSx={{ gridArea: 'response' }}>{session.response}</Card>
          <Card innerSx={{ gridArea: 'console' }}>{session.console}</Card>
        </div>
      )}
    </main>
  );
};

export default Layout;
