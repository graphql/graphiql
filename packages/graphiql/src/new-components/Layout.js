/** @jsx jsx */
import { jsx } from 'theme-ui';

import PropTypes from 'prop-types';

const Widget = ({ children }) => <div sx={{ padding: 2 }}>{children}</div>;

const Container = ({ children, mini = false }) => (
  <div
    sx={{
      backgroundColor: !mini && 'cardBackground',
      boxShadow:
        !mini &&
        (theme =>
          `0 0 0 .1px ${theme.colors.border}, 0 1px 4px 0 ${theme.colors.border}`),
    }}>
    {children}
  </div>
);

Container.propTypes = {
  mini: PropTypes.bool,
};

const Layout = ({ nav = 'nav' }) => {
  return (
    <main
      sx={{
        display: 'grid',
        gap: 2,
        padding: 2,
        gridAutoFlow: 'column',
        gridTemplateColumns: '6em',
        gridAutoColumns: 'minmax(30em, 1fr)',
        minHeight: '100vh',
      }}>
      <Container mini>{nav}</Container>
      <Container>
        <Widget>{'container'}</Widget>
      </Container>
      <Container>
        <Widget>{'container'}</Widget>
      </Container>
      <Container>
        <Widget>{'container'}</Widget>
      </Container>
    </main>
  );
};

Layout.propTypes = {
  nav: PropTypes.node,
};
export default Layout;
