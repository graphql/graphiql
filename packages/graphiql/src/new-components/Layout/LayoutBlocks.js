/** @jsx jsx */
import { jsx } from 'theme-ui';

import PropTypes from 'prop-types';

const CardRowText = ({ children }) => (
  <p
    sx={{
      padding: 2,
    }}>
    {children}
  </p>
);

const CardRow = ({ children, flex = false }) => (
  <div
    sx={{
      overflow: 'auto',
      flex: flex && '1 1 auto',
    }}>
    {children}
  </div>
);
CardRow.propTypes = {
  flex: PropTypes.bool,
};

const Card = ({ children }) => (
  <div
    sx={{
      backgroundColor: 'cardBackground',
      display: 'flex',
      flexDirection: 'column',
      overflow: 'hidden',
      boxShadow: theme =>
        `0 0 0 .1px ${theme.colors.border}, 0 1px 4px 0 ${theme.colors.border}`,
      '> *:not(:first-child)': {
        borderTop: theme => `1px solid ${theme.colors.border}`,
      },
      '> *': {
        flex: '0 0 auto',
      },
    }}>
    {children}
  </div>
);

Card.propTypes = {
  mini: PropTypes.bool,
};

export { Card, CardRow, CardRowText };
