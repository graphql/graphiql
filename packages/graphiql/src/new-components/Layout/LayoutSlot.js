/** @jsx jsx */
import { jsx } from 'theme-ui';
import PropTypes from 'prop-types';

const LayoutSlot = ({ children, name }) => (
  <div
    sx={{
      display: 'grid',
      gridTemplate: '100% / 100%',
    }}
    data-slot={name}>
    {children}
  </div>
);

LayoutSlot.propTypes = { name: PropTypes.string };
export default LayoutSlot;
