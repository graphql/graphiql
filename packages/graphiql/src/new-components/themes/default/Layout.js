/** @jsx jsx */
import { jsx } from 'theme-ui';

import PropTypes from 'prop-types';
import { PANEL_SIZES, LAYOUT_PROP_TYPES } from './../constants';

const NAV_WIDTH = '6em';
const CONTENT_MIN_WIDTH = '60em';

const sizeInCSSUnits = (theme, size) => {
  switch (size) {
    case 'sidebar':
      return '10em';
    case 'aside':
      return '20em';
    default:
      return `calc(100vw - ${theme.space[2] * 3}px - ${NAV_WIDTH})`;
  }
};

const Card = ({ children, size, transparent = false, innerSx }) => (
  <div
    sx={{
      display: 'grid',
      backgroundColor: !transparent && 'cardBackground',
      boxShadow: !transparent && 'card',
      minWidth: size && (theme => sizeInCSSUnits(theme, size)),
      gridTemplate: '100% / 100%',
      ...innerSx,
    }}>
    {children}
  </div>
);

Card.propTypes = {
  transparent: PropTypes.bool,
  size: PropTypes.oneOf(PANEL_SIZES),
  innerSx: PropTypes.object,
};

const gridBase = {
  display: 'grid',
  gridAutoFlow: 'column',
  gridAutoColumns: '1fr',
  gridAutoRows: '100%',
  gap: 2,
};

const Layout = ({ nav, navPanels, explorer }) => {
  const hasNavPanels = navPanels && navPanels.length > 0;
  return (
    <main
      sx={{
        ...gridBase,
        padding: 2,
        gridTemplate: hasNavPanels
          ? `'nav panels explorer' 100% / ${NAV_WIDTH} min-content minmax(${CONTENT_MIN_WIDTH}, 1fr)`
          : `'nav explorer' 100% / ${NAV_WIDTH} minmax(${CONTENT_MIN_WIDTH}, 1fr)`,
        height: '100vh',
      }}>
      {nav && (
        <Card innerSx={{ gridArea: 'nav' }} transparent>
          {nav}
        </Card>
      )}
      {hasNavPanels && (
        <div
          sx={{
            gridArea: 'panels',
            ...gridBase,
          }}>
          {navPanels.map(({ component, key, size }) => (
            <Card key={key} size={size}>
              {component}
            </Card>
          ))}
        </div>
      )}
      {explorer && (
        <div
          sx={{
            ...gridBase,
            gridArea: 'explorer',
            gridAutoRows: '1fr',
            gridTemplateAreas: `'input response' 'console console'`,
          }}>
          <Card>{explorer.input}</Card>
          <Card>{explorer.response}</Card>
          <Card innerSx={{ gridArea: 'console' }}>{explorer.console}</Card>
        </div>
      )}
    </main>
  );
};

Layout.propTypes = LAYOUT_PROP_TYPES;

export default Layout;
