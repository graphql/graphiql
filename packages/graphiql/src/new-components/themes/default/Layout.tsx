/** @jsx jsx */
import { jsx, SxStyleProp } from 'theme-ui';
import { PropsWithChildren } from 'react';
import { GraphiQLTheme, PanelSize, LayoutPropTypes } from '../types';

const NAV_WIDTH = '6em';
const CONTENT_MIN_WIDTH = '60em';

function sizeInCSSUnits(theme: GraphiQLTheme, size: PanelSize) {
  switch (size) {
    case 'sidebar':
      return '10em';
    case 'aside':
      return '20em';
    default:
      return `calc(100vw - ${theme.space[2] * 3}px - ${NAV_WIDTH})`;
  }
}

type CardPropTypes = PropsWithChildren<{
  size?: PanelSize;
  transparent?: boolean;
  innerSx?: SxStyleProp;
}>;

const Card = ({
  children,
  size,
  transparent = false,
  innerSx,
}: CardPropTypes) => (
  <div
    sx={
      {
        display: 'grid',
        backgroundColor: !transparent ? 'cardBackground' : undefined,
        boxShadow: !transparent && 'card',
        minWidth: size && (theme => sizeInCSSUnits(theme, size)),
        gridTemplate: '100% / 100%',
        ...(innerSx ?? {}),
      } as SxStyleProp
    }>
    {children}
  </div>
);

const gridBase = {
  display: 'grid',
  gridAutoFlow: 'column',
  gridAutoColumns: '1fr',
  gridAutoRows: '100%',
  gap: 3,
};

const Layout = ({ nav, navPanels, explorer }: LayoutPropTypes) => {
  const hasNavPanels = (navPanels && navPanels?.length > 0) || false;
  return (
    <main
      sx={{
        ...gridBase,
        padding: 3,
        gridTemplate: hasNavPanels
          ? `'nav panels explorer' 100% / ${NAV_WIDTH} min-content minmax(${CONTENT_MIN_WIDTH}, 1fr)`
          : `'nav explorer' 100% / ${NAV_WIDTH} minmax(${CONTENT_MIN_WIDTH}, 1fr)`,
        height: '100vh',
      }}>
      {nav && (
        <Card innerSx={{ gridArea: 'nav' }} transparent>
          {nav}
        </Card>
      )}
      {hasNavPanels && (
        <div
          sx={{
            gridArea: 'panels',
            ...gridBase,
          }}>
          {navPanels!.map(({ component, key, size }) => (
            <Card key={key} size={size}>
              {component}
            </Card>
          ))}
        </div>
      )}
      {explorer && (
        <div
          sx={{
            ...gridBase,
            gridArea: 'explorer',
            gridAutoRows: '1fr',
            gridTemplateAreas: `'input response' 'console console'`,
          }}>
          <Card>{explorer.input}</Card>
          <Card>{explorer.response}</Card>
          <Card innerSx={{ gridArea: 'console' }}>{explorer.console}</Card>
        </div>
      )}
    </main>
  );
};

export default Layout;
