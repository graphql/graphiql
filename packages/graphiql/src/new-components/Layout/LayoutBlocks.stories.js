import React from 'react';
import { Card, CardRow } from './LayoutBlocks';

export default { title: 'Layout Cards' };

export const withFlexChild = () => (
  <div style={{ height: '50vh' }}>
    <Card>
      <CardRow>{'Title'}</CardRow>
      <CardRow flex>
        {'This child is looooong'}
        {Array(300)
          .fill('scroll')
          .map((c, i) => (
            <div key={i}>{c}</div>
          ))}
      </CardRow>
    </Card>
  </div>
);
