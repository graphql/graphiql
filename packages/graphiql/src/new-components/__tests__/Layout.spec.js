import React from 'react';
import renderer from 'react-test-renderer';
import Layout from '../Layout';
import Providers from './fixtures/Providers';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Providers>
        <Layout />
      </Providers>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
