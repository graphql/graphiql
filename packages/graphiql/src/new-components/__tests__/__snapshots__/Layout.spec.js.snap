// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  className="css-288br5-Layout"
>
  <div
    className="css-445rez-Container"
  >
    nav
  </div>
  <div
    className="css-99y0y9-Container"
  >
    <div
      className="css-dj446p-Widget"
    >
      container
    </div>
  </div>
  <div
    className="css-99y0y9-Container"
  >
    <div
      className="css-dj446p-Widget"
    >
      container
    </div>
  </div>
  <div
    className="css-99y0y9-Container"
  >
    <div
      className="css-dj446p-Widget"
    >
      container
    </div>
  </div>
</main>
`;
